core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala

```diff
@@ -443,9 +466,27 @@ private[spark] class ExecutorAllocationManager(
   private def onExecutorIdle(executorId: String): Unit = synchronized {
     if (executorIds.contains(executorId)) {
       if (!removeTimes.contains(executorId) && !executorsPendingToRemove.contains(executorId)) {
+
+        val hasCachedBlocks =
+          executorsWithCachedBlocks.contains(executorId) ||
+            executorEndpoints.get(executorId).exists(_.askWithRetry[Boolean](HasCachedBlocks))
+
+        if (hasCachedBlocks) executorsWithCachedBlocks += executorId
```

Comment:
by the way it's OK to keep the map, but I wonder if it actually does much in practice because in many cases we'll do this exactly once for each executor.